storage/bootstrap/bootstrapper/commitlog/source.go

```diff
@@ -96,56 +111,132 @@ func (s *commitLogSource) Read(
-		bopts       = s.opts.ResultOptions()
-		blopts      = bopts.DatabaseBlockOptions()
-		blockSize   = ns.Options().RetentionOptions().BlockSize()
-		encoderPool = bopts.DatabaseBlockOptions().EncoderPool()
-		errs        = 0
+		namespace    = ns.ID()
+		numShards    = len(shardsTimeRanges)
+		unmerged     = []map[ts.Hash]encodersByTime{}
+		unmergedLock = &sync.RWMutex{}
+		numConc      = s.opts.M3TSZEncodingConcurrency()
+		bopts        = s.opts.ResultOptions()
+		blopts       = bopts.DatabaseBlockOptions()
+		blockSize    = ns.Options().RetentionOptions().BlockSize()
+		encoderPool  = bopts.DatabaseBlockOptions().EncoderPool()
+		workerErrs   = make([]int, numConc)
 	)
+
+	encoderChans := []chan encoderArg{}
+	for i := 0; i < numConc; i++ {
+		encoderChans = append(encoderChans, make(chan encoderArg, encoderChanBufSize))
+	}
+
+	// Spin up numConc background go-routines to handle M3TSZ encoding. This must
+	// happen before we start reading to prevent infinitely blocking writes to
+	// the encoderChans.
+	wg := &sync.WaitGroup{}
+	for workerNum, encoderChan := range encoderChans {
+		wg.Add(1)
+		go s.startM3TSZEncodingWorker(
+			workerNum,
+			encoderChan,
+			&unmerged,
+			unmergedLock,
+			encoderPool,
+			workerErrs,
+			blopts,
+			wg,
+		)
+	}
+
 	for iter.Next() {
 		series, dp, unit, annotation := iter.Current()
-
-		// check if the series belongs to current namespace being bootstrapped
-		if !namespace.Equal(series.Namespace) {
+		if !s.shouldEncodeSeries(namespace, shardsTimeRanges, blockSize, series, dp) {
 			continue
 		}
 
-		ranges, ok := shardsTimeRanges[series.Shard]
-		if !ok {
-			// Not bootstrapping this shard
-			continue
+		// We use the unmerged slice for lookup because its faster than a map.
+		// Shard ID's map directly to indices, but we don't know the highest shard
+		// number upfront so we resize when necessary. We don't need to acquire a
+		// lock when we check the length because this is the only goroutine that
+		// change the size of the slice.
+		if len(unmerged) < int(series.Shard+1) {
+			// Resize shard
+			unmergedLock.Lock()
+			oldUnmerged := unmerged
+			unmerged = make([]map[ts.Hash]encodersByTime, series.Shard+1)
+			copy(unmerged, oldUnmerged)
+			unmergedLock.Unlock()
 		}
 
-		blockStart := dp.Timestamp.Truncate(blockSize)
-		blockEnd := blockStart.Add(blockSize)
-		blockRange := xtime.Range{
-			Start: blockStart,
-			End:   blockEnd,
-		}
-		if !ranges.Overlaps(blockRange) {
-			// Data in this block does not match the requested ranges
-			continue
+		// Distribute work such that each encoder goroutine is responsible for
+		// approximately numShards / numConc shards. This also means that all
+		// datapoints for a given shard/series will be processed in a serialized
+		// manner.
+		encoderChans[series.Shard%uint32(numConc)] <- struct {
+			series     commitlog.Series
+			dp         ts.Datapoint
+			unit       xtime.Unit
+			annotation ts.Annotation
+			blockStart time.Time
+		}{
```

Comment:
Better to actually used the declared type that inline declare a type, e.g. `<- encoderArg{ /* fields */ }`

Reply:
Yeah good call, I have no idea why I did that